contracts/cw-shuffle-voting/src/contract.rs

```diff
@@ -0,0 +1,362 @@
+            };
+            let msg = SubMsg::reply_on_success(msg, INSTANTIATE_TOKEN_REPLY_ID);
+
+            Ok(Response::default()
+                .add_attribute("action", "instantiate")
+                .add_attribute("token", "new_token")
+                .add_submessage(msg))
+        }
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::StakeChangedHook(stake_changed_hook_msg) => {
+            execute_stake_changed_hook(deps, env, info, stake_changed_hook_msg)?;
+        }
+    }
+    Ok(Response::new().add_attribute("action", "execute"))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::VotingPowerAtHeight { address, height } => {
+            query_voting_power_at_height(deps, env, address, height)
+        }
+        QueryMsg::TotalPowerAtHeight { height } => query_total_power_at_height(deps, env, height),
+        QueryMsg::Info {} => query_info(deps),
+    }
+}
+
+pub fn query_voting_power_at_height(
+    deps: Deps,
+    env: Env,
+    address: String,
+    height: Option<u64>,
+) -> StdResult<Binary> {
+    let address = deps.api.addr_validate(&address)?;
+    let power = VOTE_WEIGHTS
+        .may_load_at_height(deps.storage, &address, height.unwrap_or(env.block.height))?
+        .unwrap_or_default();
+
+    to_binary(&cw_core_interface::voting::VotingPowerAtHeightResponse {
+        power,
+        height: env.block.height,
+    })
+}
+
+pub fn execute_stake_changed_hook(
+    deps: DepsMut,
+    env: Env,
+    _info: MessageInfo,
+    stake_changed_hook_msg: StakeChangedHookMsg,
+) -> Result<(), ContractError> {
+    match stake_changed_hook_msg {
+        StakeChangedHookMsg::Stake {
+            addr,
+            amount: _,
+            staked_addresses_count,
+        } => {
+            shuffle_voting_power(deps, env, staked_addresses_count, Some(addr), None)?;
+        }
+        StakeChangedHookMsg::Unstake {
+            addr,
+            amount: _,
+            staked_addresses_count,
+        } => {
+            shuffle_voting_power(deps, env, staked_addresses_count, None, Some(addr))?;
+        }
+    }
+    Ok(())
+}
+
+// Randomly distributes vote power amongst addresses.
+pub fn shuffle_voting_power(
+    deps: DepsMut,
+    env: Env,
+    addresses_count: u64,
+    addr_to_add: Option<Addr>,
+    addr_to_remove: Option<Addr>,
+) -> StdResult<()> {
+    // Generate voting weights based on total staked amount and number of participating addresses.
+    let weights = generate_random_voting_weights(env.block.height, addresses_count);
+
+    // Remove unstaked address if exists
+    if let Some(addr) = addr_to_remove {
+        VOTE_WEIGHTS.remove(deps.storage, &addr, env.block.height)?;
+    }
```

Comment:
What happens here if someone partially unstakes? Ex: I have 50 tokens staked and unstake 25.